### PR TITLE
Create blog post for "Type-Based Alias Analysis"

### PR DESCRIPTION
**Pull request by zhouyuan1119**

### CODE REVIEW

File: content/blog/2019-09-29-tbaa.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+- Dynamic evaluation actually examines how the alias analysis affects the runtime of the optimized program when used together
+with other optimization passes. It actually reflects how effective the alias analysis assists the optimizations. On the downside, 
+results of dynamic evaluation depends on the benchmarks, inputs, and the "client optimization" that uses alias analysis. In 
+this paper, Redundant Load Elimination (RLE) is implemented as this "client". 
+- Limit analysis evaluates how much improvement we can possibly get compared with a "perfect" alias analysis. It can be performed
+together with dynamic evaluation, where the oracle alias pairs can be found by profiling. 
+
+### Experimental Setup
+The authors assembled their own benchmark suite to evaluate the performance of TBAA. This is interesting since the standard
+SPEC benchmarks came out in 1992 and got revised in 1995. I am not too familiar with the benchmarks so I can not comment on
```
Comment (sampsyo) [inline content/blog/2019-09-29-tbaa.md]: Maybe the problem is just that they're not written in Modula-3?

File: content/blog/2019-09-29-tbaa.md [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+RLE simplifies redundant memory expressions with variable references, and tries to move the memory references out of the loop
+if there is no alias inside the loop. Similar optimizations also apply to branches. 
+
+#### Results
+By statically examining the number of removed redundant loads, we can draw the same conclusion as in the static evaluation 
+section: *FieldTypeDecl* and *SMFieldTypeRefs* are strictly more powerful than *TypeDecl*. However, when comparing the 
+performance of optimized programs, all three techniques offer similar speedup when used together with RLE. This result 
+demonstrates that at least for RLE, a more precise alias analysis may not provide much benefit over TBAA. The limit analysis
+confirms this statement, because almost all redundant loads can be removed by using RLE together with TBAA for most benchmarks. 
+For the redundant loads that can not be removed, the authors studied the cause and found that they are mostly caused by problems
```
Comment (sampsyo) [inline content/blog/2019-09-29-tbaa.md]:
```suggestion
For the redundant loads that cannot be removed, the authors studied the cause and found that they are mostly caused by problems
```

File: content/blog/2019-09-29-tbaa.md [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+- When merging types in *SMFieldTypeRefs*, any two types with a subtype relationship are merged together, since the 
+unavailable code might assign them. To my understanding, this modification makes *SMFieldTypeRefs* almost equivalent to 
+*FieldTypeDecl*. 
+
+While these changes seem very conservative, actually they do not affect the performance of TBAA that much. The authors show 
+that the "open world" assumption has negligible effect on the execution time of the compiled program when RLE is applied. The 
+result is not surprising, because even in the "close world" scenario, *SMFieldTypeRefs* does not have clear advantage over
+*FieldTypeDecl*. The modified version of *SMFieldTypeRefs* is very close to *FieldTypeDecl*, with just slightly more 
+restrictions. However, it is not mentioned in the paper **how much** of the program is still available in the "open world" 
+case, which makes the result less convincing. TBAA with only one function available and with all but one functions available
```
Comment (sampsyo) [inline content/blog/2019-09-29-tbaa.md]: While I don't know for sure, I assume the answer to "how much" of the code is available is: all of it. That is, the difference is not in how much code the analysis gets to see but in whether it needs to assume that there *might* be additional code out there that it doesn't get to observe.

Comment (sampsyo) [inline content/blog/2019-09-29-tbaa.md]:
```suggestion
case, which makes the result less convincing. TBAA with only one function available and with all but one function available
```

**Review by sampsyo — COMMENTED** [review]
Fantastic work, Yuan! This is an extremely clear writeup. Here are just a few minor suggestions and thoughts.

**Comment by zhouyuan1119** [comment]
Hi Adrian,

Thanks for the suggestions. I made some changes and added several lines to discuss
- How TBAA is still useful in current compiler frameworks;
- How TBAA may not suffice for new "clients" that come with software/hardware co-design.

**Comment by sampsyo** [comment]
Excellent! Looks great, Yuan.

**Pull request merged**